lib/connections/service.go

```diff
@@ -240,6 +232,25 @@ next:
 	}
 }
 
+// If rate limiting is set, and based on the address we
+// limit the connection, then we wrap it in a limiter.
+func (s *Service) setupConnection(c IntermediateConnection, writeRateLimit, readRateLimit *ratelimit.Bucket, options config.OptionsConfiguration) (io.Reader, io.Writer) {
+	limit := s.shouldLimit(c.RemoteAddr())
```

Comment:
It looks like the read and write limiters already correctly handle nil buckets (not doing any limiting then), so this could be simplified down to just

```go
// ...
    rd, wr := io.Reader(c), io.Writer(c)
    if s.shouldLimit(c.RemoteAddr()) {
        rd = NewReadLimiter(rd, s.readRateLimit)
        wr = NewWriteLimiter(wr, s.readRateLimit)
    }
```

at the call site, obviating the setupConnection function completely, I think?